pkg/reconciler/workspaceindex/server.go

```diff
@@ -0,0 +1,164 @@
+		if !s.stable() {
+			return errors.New("work queues have outstanding keys")
+		}
+		return nil
+	}), healthz.NewInformerSyncHealthz(s.waiter))
+	httpServer := http.Server{Addr: ":" + strconv.Itoa(s.port), Handler: mux}
+	go func() {
+		<-ctx.Done()
+		if err := httpServer.Shutdown(context.Background()); err != nil {
+			klog.Error(err)
+		}
+	}()
+	if err := httpServer.ListenAndServe(); err != nil {
+		klog.Error(err)
+	}
+}
+
+const (
+	clusterNameQuery     = "clusterName"
+	resourceVersionQuery = "resourceVersion"
+)
+
+func (s *server) handleShard(w http.ResponseWriter, r *http.Request) {
+	// ensure the cache is stable before serving
+	if !s.stable() {
+		w.Header().Set("Retry-After", "10")
+		http.Error(w, "workspace index cache is not stable", http.StatusServiceUnavailable)
+		return
+	}
+	// first, validate the input from the user
+	q := r.URL.Query()
+	clusterName := q.Get(clusterNameQuery)
+	if clusterName == "" {
+		http.Error(w, "clusterName query must not be empty", http.StatusBadRequest)
+		return
+	}
+	organization, workspace, err := helper.ParseLogicalClusterName(clusterName)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("clusterName query invalid: %v", err), http.StatusBadRequest)
+		return
+	}
+	resourceVersionString := q.Get(resourceVersionQuery)
+	if resourceVersionString == "" {
+		resourceVersionString = "0"
+	}
+	resourceVersion, err := strconv.ParseInt(resourceVersionString, 10, 64)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("resourceVersion query must be an integer: %v", err), http.StatusBadRequest)
+		return
+	}
+
+	// then, find the shard history for the workspace in question
+	history, err := s.index.Get(organization, workspace)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusNotFound)
+		return
+	}
+	if len(history) == 0 {
+		http.Error(w, fmt.Sprintf("cluster %q has no shard history", clusterName), http.StatusInternalServerError)
+		return
+	}
+
+	// finally, find the part of the history that the user asked for. We assume that the
+	// history is sorted in ascending order of resourceVersion and that no resourceVersion
+	// ranges overlap
+	var shardName string
+	if resourceVersion == 0 {
+		// if the user asked for no particular resource version or literally "0", we want
+		// up-to-date results, served from the current shard this workspace sits on
+		shardName = history[len(history)-1].Name
+	} else {
+		idx := sort.Search(len(history), func(i int) bool {
+			return history[i].LiveAfterResourceVersion < resourceVersion && history[i].LiveBeforeResourceVersion >= resourceVersion
+		})
+		if idx == len(history) {
+			// this should never happen, since the current shard should have no liveBefore set
+			http.Error(w, fmt.Sprintf("cluster %q has no shard history containing resourceVersion %d", clusterName, resourceVersion), http.StatusInternalServerError)
+			return
+		}
+		shardName = history[idx].Name
+	}
+	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
+	fmt.Fprint(w, shardName)
```

Comment:
Do we want to return json here?

Reply:
I don't think necessarily that we need this API to even exist in the future, but it's a convenient way to test this in a black-box e2e.

Reply:
so in other words: the actual proxy handler is still missing. It would take `/clusters/<org>_<worksapce>` to the very same path on the shard, i.e. `io.Copy` the connection. Would like to see this being implemented. Seems to be a slight variation of the upper code.